### PR TITLE
[WIP] WatchfaceComplication basic implementation

### PR DESCRIPTION
**Pull request by gotoextreme**
### Description of Change ###

Implement Complication, DesignElements, ComplicationProvider basic API.

Complication <--> ComplicationProvider
Complication, DesignElement <--> watchface-editor

### CODE REVIEW

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications.WatchfaceComplication.csproj [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (WonyoungChoi) [inline src/Tizen.Applications.WatchfaceComplication/Tizen.Applications.WatchfaceComplication.csproj]: If you want to include this module in Tizen.NET package, the following information is unnecessary.

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications.WatchfaceComplication.csproj [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    <Version>1.0.0</Version>
+    <Description>Watchface Complication</Description>
+    <PackageProjectUrl>https://www.tizen.org/</PackageProjectUrl>
+    <PackageLicenseUrl>https://www.apache.org/licenses/LICENSE-2.0</PackageLicenseUrl>
+    <PackageIconUrl>https://developer.tizen.org/sites/default/files/images/tizen-pinwheel-on-light-rgb_64_64.png</PackageIconUrl>
+    </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Tizen.NET" Version="4.0.0" />
```
Comment (WonyoungChoi) [inline src/Tizen.Applications.WatchfaceComplication/Tizen.Applications.WatchfaceComplication.csproj]: Use ProjectReference instead.

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications.WatchfaceComplication.sln [hunk 3]
```diff
@@ -0,0 +1,25 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 15
+VisualStudioVersion = 15.0.26730.16
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Tizen.Applications.WatchfaceComplication", "Tizen.Applications.WatchfaceComplication.csproj", "{F1EF8A92-02E2-483D-9D68-234640648CE3}"
```
Comment (WonyoungChoi) [inline src/Tizen.Applications.WatchfaceComplication/Tizen.Applications.WatchfaceComplication.sln]: Maybe you should add Tizen, Tizen.Log, Tizen.Applications.Common projects to this solution file to build it also.

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            }
+            _updatedCallback = new Interop.WatchfaceComplication.ComplicationUpdatedCallback(ComplicationUpdate);
+            ret = Interop.WatchfaceComplication.AddUpdatedCallback(_handle, _updatedCallback, IntPtr.Zero);
+            if (ret != ComplicationError.None)
+            {
+                ErrorFactory.ThrowException(ret, "Fail to add update callback");
+            }
+        }
+
+        internal IntPtr Raw
```
Comment (WonyoungChoi) [inline src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs]: you can make it more simlply
internal IntPtr Raw => _handle;

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+        public int SupportTypes
+        {
+            get
+            {
+                return _supportTypes;
+            }
+        }
+
+        Geometry IEditable.Geometry
```
Comment (WonyoungChoi) [inline src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs]: is this private property? put the private keyword if true.

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+            set
+            {
+                _editableId = value;
+            }
+        }
+        string IEditable.Name
+        {
+            get
+            {
+                return null;
```
Comment (WonyoungChoi) [inline src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs]: ??

**Comment by WonyoungChoi** [comment]
@gotoextreme Do you still need this PR? If not, please close it.

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/EditablesContainer.cs [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        /// </summary>
+        /// <param name="editableId">The target editable Id.</param>
+        /// <returns>true if the editable is already exists in edit list, ortherwise false</returns>
+        /// <since_tizen> 5 </since_tizen>
+        public bool IsExist(int editableId)
+        {
+            return (GetEditable(editableId) != null);
+        }
+
+        private void EditReady(IntPtr container, string editorAppid, IntPtr userData)
```
[the remark this answers is not in the document]
Comment (gotoextreme) [reply src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/EditablesContainer.cs]: It's method for the native callback function.
We need intptr to matching native callback signature.

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/EditablesContainer.cs [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        {
+            OnEditReady(editorAppid);
+        }
+
+
+        private void EditableUpdatedCallback(IntPtr handle, int selectedIdx,
+            int state, IntPtr userData)
```
[the remark this answers is not in the document]
Comment (gotoextreme) [reply src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/EditablesContainer.cs]: It's method for the native callback function.
We need intptr to matching native callback signature.

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs [hunk 9]
```diff
@@ -0,0 +1,625 @@
...
+            SafeBundleHandle bundleHandle;
+            ComplicationError err = Interop.WatchfaceComplication.GetCurrentData(_handle, out bundleHandle);
+            if (err != ComplicationError.None)
+                ErrorFactory.ThrowException(err, "Can not get current data");
+            Bundle data = new Bundle(bundleHandle);
+            return data;
+        }
+
+        private void ComplicationUpdatedCallback(int complicationId,
+            string providerId, ComplicationType type, IntPtr data, IntPtr userData)
```
[the remark this answers is not in the document]
Comment (gotoextreme) [reply src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs]: It's method for the native callback function.
We need intptr to matching native callback signature.

File: src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+        }
+
+        /// <summary>
+        /// Overrides this method to handle the behavior when the complication update event comes.
+        /// </summary>
+        /// <param name="providerId">The updated provider's ID.</param>
+        /// <param name="type">The updated type.</param>
+        /// <param name="data">The updated data.</param>
+        /// <since_tizen> 5 </since_tizen>
+        protected virtual void OnComplicationUpdated(string providerId, ComplicationType type, Bundle data)
```
Comment (jusung-son) [inline src/Tizen.Applications.WatchfaceComplication/Tizen.Applications/Complication.cs]: How about changing abstract method?

**Comment by gotoextreme** [comment]
temporary close

**Pull request closed without merging**